From 39f6b27c418c112a6e86865f431ccf6bfba89882 Mon Sep 17 00:00:00 2001
From: Carlos Garnacho <carlosg@gnome.org>
Date: Mon, 7 Jul 2025 17:32:47 +0200
Subject: [PATCH 3/4] print-notifications: Copy method from GTK to get paper
 size

This is just some nl_langinfo glue, we do not need all of GTK for
this.
---
 meson.build                               |  7 ++++
 plugins/print-notifications/gsd-printer.c | 41 ++++++++++++++++++++---
 plugins/print-notifications/main.c        |  4 ---
 plugins/print-notifications/meson.build   |  1 -
 4 files changed, 43 insertions(+), 10 deletions(-)

diff --git a/meson.build b/meson.build
index 671afcf8..61da83d3 100644
--- a/meson.build
+++ b/meson.build
@@ -189,6 +189,13 @@ if enable_cups
     compile_args: cups_cflags
   )
   '''
+
+  if cc.has_header('langinfo.h')
+    foreach nl_enum: [ '_NL_PAPER_HEIGHT',
+                       '_NL_PAPER_WIDTH' ]
+      config_h.set('HAVE_' + nl_enum, cc.has_header_symbol('langinfo.h', nl_enum))
+    endforeach
+  endif
 endif
 
 # Rfkill
diff --git a/plugins/print-notifications/gsd-printer.c b/plugins/print-notifications/gsd-printer.c
index 596a20a7..f1770c9f 100644
--- a/plugins/print-notifications/gsd-printer.c
+++ b/plugins/print-notifications/gsd-printer.c
@@ -23,10 +23,13 @@
 #include <libnotify/notify.h>
 #include <glib/gi18n.h>
 #include <glib/gstdio.h>
-#include <gtk/gtk.h>
 #include <cups/cups.h>
 #include <cups/ppd.h>
 
+#if defined(HAVE__NL_PAPER_HEIGHT) && defined(HAVE__NL_PAPER_WIDTH)
+#include <langinfo.h>
+#endif
+
 /* FIXME: https://gitlab.gnome.org/GNOME/gnome-settings-daemon/-/issues/860 */
 G_GNUC_BEGIN_IGNORE_DEPRECATIONS
 
@@ -744,14 +747,42 @@ printer_autoconfigure (gchar *printer_name)
         g_free (commands_lowercase);
 }
 
+#define NL_PAPER_GET(x)         \
+        ((union { char *string; unsigned int word; }) nl_langinfo(x)).word
+
 /* Returns default page size for current locale */
 static const gchar *
 get_page_size_from_locale (void)
 {
-  if (g_str_equal (gtk_paper_size_get_default (), GTK_PAPER_NAME_LETTER))
-    return "Letter";
-  else
-    return "A4";
+        char *locale;
+#if defined(HAVE__NL_PAPER_HEIGHT) && defined(HAVE__NL_PAPER_WIDTH)
+        int width = NL_PAPER_GET (_NL_PAPER_WIDTH);
+        int height = NL_PAPER_GET (_NL_PAPER_HEIGHT);
+
+        if (width == 210 && height == 297)
+                return "A4";
+
+        if (width == 216 && height == 279)
+                return "Letter";
+#endif
+
+#if defined(LC_PAPER)
+        locale = setlocale (LC_PAPER, NULL);
+#else
+        locale = setlocale (LC_MESSAGES, NULL);
+#endif
+
+        if (!locale)
+                return "A4";
+
+        /* CLDR 1.8.1
+         * http://unicode.org/repos/cldr-tmp/trunk/diff/supplemental/territory_language_information.html
+         */
+        if (g_regex_match_simple ("[^_.@]{2,3}_(BZ|CA|CL|CO|CR|GT|MX|NI|PA|PH|PR|SV|US|VE)",
+                                  locale, G_REGEX_ANCHORED, G_REGEX_MATCH_ANCHORED))
+                return "Letter";
+        else
+                return "A4";
 }
 
 static void
diff --git a/plugins/print-notifications/main.c b/plugins/print-notifications/main.c
index 5574bc56..6428efd7 100644
--- a/plugins/print-notifications/main.c
+++ b/plugins/print-notifications/main.c
@@ -1,12 +1,8 @@
-#include <gtk/gtk.h>
-
 #include "gsd-main-helper.h"
 #include "gsd-print-notifications-manager.h"
 
 int
 main (int argc, char **argv)
 {
-        gtk_init ();
-
         return gsd_main_helper (GSD_TYPE_PRINT_NOTIFICATIONS_MANAGER, argc, argv);
 }
diff --git a/plugins/print-notifications/meson.build b/plugins/print-notifications/meson.build
index 9137d5c4..58e5103e 100644
--- a/plugins/print-notifications/meson.build
+++ b/plugins/print-notifications/meson.build
@@ -6,7 +6,6 @@ sources += main_helper_sources
 
 deps = plugins_deps + [
   cups_dep,
-  gtk_dep,
   libnotify_dep
 ]
 
-- 
2.49.0

